File: rust/src/grasp.rs
// SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
// SPDX-License-Identifier: MIT

use crate::cache::EvaluationCache;
use crate::helpers::{clamp, normalize_integer_tail, safe_evaluate};
use rand::Rng;
use rand_chacha::ChaCha8Rng;
use std::time::Instant;

/// Compute adaptive alpha via linear interpolation.
pub(crate) fn get_current_alpha(
    iter_idx: usize,
    max_iterations: usize,
    alpha_min: f64,
    alpha_max: f64,
    adaptive: bool,
    alpha: f64,
) -> f64 {
    if !adaptive {
        return alpha;
    }
    let progress = iter_idx as f64 / max_iterations.max(1) as f64;
    alpha_min + (alpha_max - alpha_min) * progress
}

/// Evaluate with optional cache.
pub(crate) fn evaluate_with_cache<F>(
    candidate: &[f64],
    func: &F,
    cache: &mut Option<EvaluationCache>,
    half: usize,
) -> f64
where
    F: Fn(&[f64]) -> f64,
{
    if let Some(ref mut c) = cache {
        if let Some(v) = c.get(candidate, half) {
            return v;
        }
        let cost = safe_evaluate(func, candidate);
        c.put(candidate, half, cost);
        cost
    } else {
        safe_evaluate(func, candidate)
    }
}

/// Select an index from the Restricted Candidate List.
fn select_from_rcl(costs: &[f64], alpha: f64, rng: &mut ChaCha8Rng) -> Option<usize> {
    if costs.is_empty() {
        return None;
    }
    let min_cost = costs.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_cost = costs.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let threshold = min_cost + alpha * (max_cost - min_cost);

    let candidates: Vec<usize> = costs
        .iter()
        .enumerate()
        .filter(|(_, &c)| c <= threshold)
        .map(|(i, _)| i)
        .collect();

    if candidates.is_empty() {
        return Some(0);
    }
    let idx = rng.random_range(0..candidates.len());
    Some(candidates[idx])
}

/// Sample an integer uniformly from bounds.
fn sample_integer_from_bounds(lo: f64, hi: f64, rng: &mut ChaCha8Rng) -> f64 {
    let lo_i = lo.ceil() as i64;
    let hi_i = hi.floor() as i64;
    if lo_i > hi_i {
        return ((lo + hi) / 2.0).round();
    }
    rng.random_range(lo_i..=hi_i) as f64
}

/// Build a purely random candidate.
fn build_random_candidate(
    num_vars: usize,
    half: usize,
    lower: &[f64],
    upper: &[f64],
    rng: &mut ChaCha8Rng,
) -> Vec<f64> {
    let mut sol = vec![0.0; num_vars];
    for i in 0..half {
        sol[i] = rng.random_range(lower[i]..=upper[i]);
    }
    for i in half..num_vars {
        sol[i] = sample_integer_from_bounds(lower[i], upper[i], rng);
    }
    sol
}

/// Build a heuristic candidate (mid ± 15% noise).
fn build_heuristic_candidate(
    num_vars: usize,
    half: usize,
    lower: &[f64],
    upper: &[f64],
    rng: &mut ChaCha8Rng,
) -> Vec<f64> {
    let mut sol = vec![0.0; num_vars];
    for i in 0..half {
        let mid = (lower[i] + upper[i]) / 2.0;
        let span = upper[i] - lower[i];
        let noise = rng.random_range(-0.15..=0.15) * span;
        sol[i] = clamp(mid + noise, lower[i], upper[i]);
    }
    for i in half..num_vars {
        sol[i] = sample_integer_from_bounds(lower[i], upper[i], rng);
    }
    sol
}

/// GRASP construction phase.
#[allow(clippy::too_many_arguments)]
pub(crate) fn construct_grasp<F>(
    num_vars: usize,
    lower: &[f64],
    upper: &[f64],
    func: &F,
    initial_guess: Option<&[f64]>,
    alpha: f64,
    half: usize,
    num_candidates: usize,
    cache: &mut Option<EvaluationCache>,
    rng: &mut ChaCha8Rng,
    deadline: Option<Instant>,
) -> (Vec<f64>, f64)
where
    F: Fn(&[f64]) -> f64,
{
    let mut candidates: Vec<Vec<f64>> = Vec::with_capacity(num_candidates);
    let mut costs: Vec<f64> = Vec::with_capacity(num_candidates);

    // Initial guess as first candidate
    if let Some(ig) = initial_guess {
        let mut sol = ig.to_vec();
        normalize_integer_tail(&mut sol, half);
        let cost = evaluate_with_cache(&sol, func, cache, half);
        candidates.push(sol);
        costs.push(cost);
    }

    // One heuristic candidate
    if candidates.len() < num_candidates {
        let mut sol = build_heuristic_candidate(num_vars, half, lower, upper, rng);
        normalize_integer_tail(&mut sol, half);
        let cost = evaluate_with_cache(&sol, func, cache, half);
        candidates.push(sol);
        costs.push(cost);
    }

    // Fill rest with random candidates
    while candidates.len() < num_candidates {
        if crate::helpers::expired(deadline) {
            break;
        }
        let mut sol = build_random_candidate(num_vars, half, lower, upper, rng);
        normalize_integer_tail(&mut sol, half);
        let cost = evaluate_with_cache(&sol, func, cache, half);
        candidates.push(sol);
        costs.push(cost);
    }

    // RCL selection — costs is always non-empty here (heuristic candidate always added)
    let idx = select_from_rcl(&costs, alpha, rng).expect("costs should not be empty");
    (candidates.swap_remove(idx), costs[idx])
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_chacha::ChaCha8Rng;
    use std::time::{Duration, Instant};

    #[test]
    fn test_select_from_rcl_empty_returns_none() {
        let mut rng = ChaCha8Rng::seed_from_u64(0);
        assert!(select_from_rcl(&[], 0.5, &mut rng).is_none());
    }

    #[test]
    fn test_construct_grasp_expired_deadline_breaks_fill_loop() {
        // An already-expired deadline triggers the break in the fill-random loop.
        // The heuristic candidate is still returned because it was added before the loop.
        let mut rng = ChaCha8Rng::seed_from_u64(42);
        let mut cache = None;
        let lower = [-5.0f64, -5.0];
        let upper = [5.0f64, 5.0];
        let func = |x: &[f64]| x.iter().map(|&xi| xi * xi).sum::<f64>();
        let deadline = Some(Instant::now() - Duration::from_secs(1));
        let (sol, cost) = construct_grasp(
            2, &lower, &upper, &func, None, 0.5, 2, 100, &mut cache, &mut rng, deadline,
        );
        assert_eq!(sol.len(), 2);
        assert!(cost.is_finite());
    }
}

File: julia/src/grasp.jl
# SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
# SPDX-License-Identifier: MIT

"""GRASP — Greedy Randomized Adaptive Search Procedure."""

"""
    validate_bounds_and_initial!(lower, upper, initial_guess, num_vars)

Validate that `lower` and `upper` have length `num_vars`, that `upper[i] > lower[i]`
for all variables, and that `initial_guess` (if provided) lies strictly within bounds.

Throws [`InvalidBoundsError`](@ref) or [`InvalidInitialGuessError`](@ref) on failure.
"""
function validate_bounds_and_initial!(
    lower::Vector{Float64},
    upper::Vector{Float64},
    initial_guess::Union{Vector{Float64}, Nothing},
    num_vars::Int,
)
    length(lower) != num_vars && throw(
        InvalidBoundsError(
            "lower (len=$(length(lower))) must have length num_vars=$num_vars",
        ),
    )
    length(upper) != num_vars && throw(
        InvalidBoundsError(
            "upper (len=$(length(upper))) must have length num_vars=$num_vars",
        ),
    )

    if initial_guess !== nothing
        length(initial_guess) != num_vars && throw(
            InvalidInitialGuessError(
                "initial_guess has length $(length(initial_guess)), expected $num_vars",
            ),
        )
        for i in 1:num_vars
            if initial_guess[i] <= lower[i] || initial_guess[i] >= upper[i]
                throw(
                    InvalidInitialGuessError(
                        "initial_guess values must be strictly between lower and upper; violating index: $i",
                    ),
                )
            end
        end
    end
end

function get_current_alpha(iter_idx::Int, config::GIVPConfig)::Float64
    if !config.adaptive_alpha
        return config.alpha
    end
    progress = iter_idx / max(1, config.max_iterations - 1)
    return config.alpha_min + (config.alpha_max - config.alpha_min) * progress
end

function evaluate_with_cache(
    cand::Vector{Float64},
    evaluator::Function,
    cache::Union{EvaluationCache, Nothing},
)::Float64
    if cache !== nothing
        cached_cost = cache_get(cache, cand)
        cached_cost !== nothing && return cached_cost
    end
    cost = safe_evaluate(evaluator, cand)
    if cache !== nothing && isfinite(cost)
        cache_put!(cache, cand, cost)
    end
    return cost
end

function select_from_rcl(
    costs::Vector{Float64},
    alpha::Float64,
    rng::AbstractRNG,
)::Union{Int, Nothing}
    valid_mask = isfinite.(costs)
    !any(valid_mask) && return nothing
    valid_idx = findall(valid_mask)
    valid_costs = costs[valid_idx]
    min_cost = minimum(valid_costs)
    max_cost = maximum(valid_costs)
    threshold = min_cost + alpha * (max_cost - min_cost)
    rcl_local = valid_idx[valid_costs .<= threshold]
    isempty(rcl_local) && (rcl_local = valid_idx)
    return rcl_local[rand(rng, 1:length(rcl_local))]
end

function normalize_integer_tail!(sol::Vector{Float64}, half::Int)
    for idx in (half + 1):length(sol)
        sol[idx] = Float64(round(Int, sol[idx]))
    end
end

function sample_integer_from_bounds(
    lower::Float64,
    upper::Float64,
    rng::AbstractRNG,
)::Float64
    lo = ceil(Int, lower)
    hi = floor(Int, upper)
    hi >= lo && return Float64(rand(rng, lo:hi))
    return Float64(round(Int, (lower + upper) / 2.0))
end

function build_random_candidate(
    num_vars::Int,
    half::Int,
    lower::Vector{Float64},
    upper::Vector{Float64},
    rng::AbstractRNG,
)::Vector{Float64}
    sol = Vector{Float64}(undef, num_vars)
    for i in 1:half
        sol[i] = lower[i] + (upper[i] - lower[i]) * rand(rng)
    end
    for i in (half + 1):num_vars
        sol[i] = sample_integer_from_bounds(lower[i], upper[i], rng)
    end
    return sol
end

function build_heuristic_candidate(
    num_vars::Int,
    half::Int,
    lower::Vector{Float64},
    upper::Vector{Float64},
    rng::AbstractRNG,
)::Vector{Float64}
    sol = Vector{Float64}(undef, num_vars)
    mid = (lower[1:half] .+ upper[1:half]) ./ 2.0
    span = upper[1:half] .- lower[1:half]
    noise = rand(rng, half) .* 0.3 .- 0.15  # uniform(-0.15, 0.15)
    sol[1:half] = clamp.(mid .+ noise .* span, lower[1:half], upper[1:half])

    for idx in (half + 1):num_vars
        lo = ceil(Int, lower[idx])
        hi = floor(Int, upper[idx])
        cont_idx = idx - half
        if hi > lo && cont_idx <= half && span[cont_idx] > 0
            frac = (sol[cont_idx] - lower[cont_idx]) / span[cont_idx]
            target = lo + frac * (hi - lo)
            sol[idx] = Float64(clamp(round(Int, target), lo, hi))
        else
            sol[idx] =
                hi >= lo ? Float64(hi) :
                Float64(round(Int, (lower[idx] + upper[idx]) / 2.0))
        end
    end
    return sol
end

function construct_grasp(
    num_vars::Int,
    lower::Vector{Float64},
    upper::Vector{Float64},
    evaluator::Function,
    initial_guess::Union{Vector{Float64}, Nothing},
    alpha::Float64;
    seed::Union{Int, Nothing} = nothing,
    num_candidates_per_step::Int = 20,
    cache::Union{EvaluationCache, Nothing} = nothing,
    n_workers::Int = 1,
)::Vector{Float64}
    rng = new_rng(seed)
    validate_bounds_and_initial!(lower, upper, initial_guess, num_vars)

    half = get_half(num_vars)
    n_candidates = max(num_candidates_per_step, 5)
    candidates = Vector{Float64}[]

    # Seed from initial guess
    if initial_guess !== nothing
        seed_candidate = copy(initial_guess)
        seed_cost = safe_evaluate(evaluator, seed_candidate)
        if !isfinite(seed_cost)
            seed_candidate = lower .+ (upper .- lower) .* rand(rng, num_vars)
        end
        normalize_integer_tail!(seed_candidate, half)
        push!(candidates, seed_candidate)
    end

    # Heuristic candidate
    heur = build_heuristic_candidate(num_vars, half, lower, upper, rng)
    normalize_integer_tail!(heur, half)
    push!(candidates, heur)

    # Random candidates
    while length(candidates) < n_candidates
        sol = build_random_candidate(num_vars, half, lower, upper, rng)
        normalize_integer_tail!(sol, half)
        push!(candidates, sol)
    end

    # Evaluate: sequential with cache (n_workers=1) or parallel without cache
    costs = Vector{Float64}(undef, length(candidates))
    if n_workers > 1
        # Parallel evaluation: bypass cache (Dict is not thread-safe in Julia)
        Threads.@threads for i in eachindex(candidates)
            costs[i] = safe_evaluate(evaluator, candidates[i])
        end
    else
        for i in eachindex(candidates)
            costs[i] = evaluate_with_cache(candidates[i], evaluator, cache)
        end
    end

    # Select from RCL
    best_idx = select_from_rcl(costs, alpha, rng)
    if best_idx === nothing
        best_idx = argmin(costs)
    end
    return copy(candidates[best_idx])
end
